### ISSUE TITLE
MutationEndpoint with return value

### ISSUE THREAD

**Issue by Paxol** [Jcparkyn/Phetch#3]
I have an api endpoint that creates records in the database, so a 'mutation', and returns the created object.
For this use cases would be nice to have a MutationEndpoint that returns something.

If I want to do this now, correct me if I'm wrong, I would need to define an `Endpoint` and set it's cache value to 0. 
There would be some side effects that I should be aware of?

**Comment by Jcparkyn** [comment]
It sounds like you might be coming from something like React Query, where "query" and "mutation" mean slightly different things than they do in Phetch. If I'm understanding your use case correctly, you should just be able to use a normal `Endpoint` but call it with `.Trigger()` (and don't provide an `Arg` to `<UseEndpoint/>`, if you're using that). This is actually the recommended way to do 90% of what would be called "mutations" in React Query.

The docs and samples could admittedly use some work here, but these two sections of the readme should hopefully explain most of it: [Mutations and Parameterless Queries](https://github.com/Jcparkyn/Phetch#mutations-and-parameterless-queries), [Invoking queries manually](https://github.com/Jcparkyn/Phetch#invoking-queries-manually).

I'm considering renaming all the "Mutation" classes to something else (maybe `ResultlessQuery`, `ResultlessEndpoint` etc?) before releasing v1.0, to hopefully remove this confusion. Realistically, these should be very rarely used (since even most PUT/POST/DELETE endpoints have some sort of response), so a more verbose name shouldn't be much of an issue.

**Comment by Paxol** [comment]
Yes, I'm coming from React Query and ended up using `.Trigger` as you mentioned.
Definitely agree on the renaming.
Excited to see the development of the library and maybe contribute if needed.

**Comment by Jcparkyn** [comment]
Nice, definitely let me know if you run into anything else unintuitive.

For a bit more context into the naming, Phetch initially had queries and mutations very similar to React Query, but along the way I realized that combining the two would remove a lot of duplicate code and fill some feature gaps in RQ (and I saw comments from RQ maintainers expressing the same sentiment). Some of the old design choices and naming stayed around, even when they probably shouldn't have.

**Comment by max-programming** [comment]
Hey @Jcparkyn 
This library is amazing! Thanks a lot for working on this

Just wanted to know if this is the right way of doing a "mutation"? I found the `.Invoke()` method so used that. How does it differentiate from `.Trigger()`

I feel like this way I can also use the `IsLoading` and `IsError` states

```razor
@code {
    public readonly LoginFormModel LoginForm = new("", "");
    
    private Query<LoginFormModel, dynamic>? _loginUserMutation;

    protected override void OnInitialized()
    {
        _loginUserMutation = Api.LoginUserEndpoint.Use();
    }

    public class LoginFormModel(string username, string password)
    {
        [Required(ErrorMessage = "Username is required")]
        public string Username { get; set; } = username;

        [Required(ErrorMessage = "Password is required")]
        public string Password { get; set; } = password;
    }

    private async void HandleSubmit()
    {
        if (_loginUserMutation is null) return;
        await _loginUserMutation.Invoke(LoginForm);
    }

}
```

**Comment by Jcparkyn** [comment]
@max-programming 

> even after I get the console output (i.e., the request is over)

In your code the `Console.WriteLine` doesn't actually occur after the request is over, only after the request has been started. If you want the console log to happen after the request completes, either do:
1. `_loginUserMutation.Trigger(LoginForm, onSuccess: (e) => {...});`
2. Make `HandleSubmit` `async` and use `await _loginUserMutation.TriggerAsync(LoginForm);`. Keep in mind that this will throw an exception if the request fails, which probably isn't what you want.

> that still remains true

I can't tell from the code you provided, but if you're using the query in the UI (which I assume you are) you'll also need to have an [`<ObserveQuery/>`](https://jcparkyn.github.io/phetch/docs/using-query-objects-directly.html) component to make sure your component re-renders whenever the query updates.

As another general pointer (but probably not what's causing the above), you should avoid using mutable classes (like your `LoginFormModel`) as query args, since Phetch stores the args and uses them to determine whether to re-fetch a query. If you mutate an object after using it as an arg, the version stored by Phetch will also be mutated (since it's the same object). This is much less of a problem for `Trigger` than regular queries, but will still cause some odd behaviour.

It might seem like a hassle, but there are lots of other good reasons to separate your API models from your form models, so generally I'd recommend just do that. Then use a function to convert from the mutable form model to the immutable API model, and use the API model as the parameter for your endpoint. `record` classes work great as API models, since they're immutable and have good implementations of `Equals` and `GetHashCode` by default.
